### PR TITLE
Fix build on Windows

### PR DESCRIPTION
**Pull request by raphinesse**
This fixes the package installation and test running on Windows to expose failing tests related to #1643.

More details in the commit messages.

### CODE REVIEW

**Comment by raphinesse** [comment]
Apparently using `nyc@15` does not support `node@<8`. Given that those Node.js versions are EOL, this does not seem to be a big issue to me. However, that fact would have to be addressed with further changes.

**Comment by ljharb** [comment]
EOL is irrelevant; we'll support what our supported eslint versions work on.

File: package.json [hunk 1]
```diff
@@ -31,6 +30,7 @@
     "test-compiled": "npm run prepublish && NODE_PATH=./lib mocha --compilers js:babel-register --recursive tests/src",
     "test-all": "npm test && for resolver in ./resolvers/*; do cd $resolver && npm test && cd ../..; done",
     "prepublish": "npm run build",
+    "prepublishOnly": "npm run copy-metafiles",
```
Comment (ljharb) [inline package.json]: why is this changed? we want it running after the build step, not just on prepublish.

File: package.json [hunk 2]
```diff
@@ -76,7 +76,7 @@
     "eslint-plugin-import": "2.x",
     "linklocal": "^2.8.2",
     "mocha": "^3.5.3",
-    "nyc": "^11.9.0",
+    "nyc": "^15.0.0",
```
Comment (ljharb) [inline package.json]: I'd prefer not to upgrade this; i've asked on https://github.com/istanbuljs/nyc/issues/1205 about a workaround.

File: package.json (hunk 1 above)
Comment (raphinesse) [reply package.json]: Why do you want it to run after the build step? From what it does, it seems to be only necessary or at all meaningful if done before packaging.

The reason for limiting  the `copy-metafiles` script to `prepublishOnly` is that it throws a syntax error if run on Windows (because it's a `sh` script), one of the two reasons the tests don't run on Appveyor CI right now.

This is a quick fix that only prevents publishing from Windows versus preventing to publish or even to do a clean `npm install`.

File: package.json (hunk 2 above)
Comment (raphinesse) [reply package.json]: Another option would be to decouple testing and coverage collection and either

- update to `nyc@15` and stop collecting coverage on node@<8
- stop collecting coverage on windows

**Comment by raphinesse** [comment]
Playing around with the node or npm version in the CI config _might_ serve to work around the issue of tests not running on Appveyor but it won't resolve the issue for someone cloning the repo to their local Windows machine.

**Comment by ljharb** [comment]
That is true. I'll keep playing with alternatives.

**Comment by ljharb** [comment]
k, looks like adding in-publish, and downgrading npm on node 12, has exposed 50 failing tests, that seem legitimate.

File: package.json (hunk 1 above)
Comment (ljharb) [reply package.json]: I fixed this by adding `in-publish` rather than using the hacky `prepublishOnly` or by breaking `npm run build`.

**Comment by raphinesse** [comment]
Regarding your changes to the `prepublish` script: Is it intended that the `build` script is neither run on local `npm install` without any arguments nor on `npm pack`? You can achieve that by using the built-in `prepublishOnly` w/out using any additional dependencies. I would consider that less hacky.

**Comment by ljharb** [comment]
`prepublishOnly` doesn't work on all npm versions, nor does `prepare`; I'll only ever use `prepublish` on any project.

**Comment by raphinesse** [comment]
> `prepublishOnly` doesn't work on all npm versions, nor does `prepare`; I'll only ever use `prepublish` on any project.

Ah, I see. against the backdrop of trying to support as many `npm` versions as possible, it makes sense, of course.


What about the other part of my question?
> Is it intended that the `build` script is neither run on local `npm install` without any arguments nor on `npm pack`?

**Comment by armano2** [comment]
there seems to be more issues on windows

https://github.com/armano2/test-eslint-monorpo/actions/runs/36409848

**Comment by ljharb** [comment]
@raphinesse no; that's what i added `not-in-publish`; that way the build script only runs with `npm publish`, and not with `npm install`.

@armano2 the current appveyor failures in this PR seem to be legitimate issues on windows, yes.

**Comment by fisker** [comment]
I think #1651 is the reason those tests fails

**Comment by ljharb** [comment]
@fisker i rebased #1651 on top of this, and its tests are still failing (albeit with 1 failure instead of 50). Almost there :-)

File: appveyor.yml [hunk 3]
```diff
@@ -22,11 +22,10 @@ install:
   # Get the latest stable version of Node.js or io.js
   - ps: Install-Product node $env:nodejs_version
 
+  # Use npm@6.10.3 to work-around issues w/ nyc@<15 on Windows
+  - npm install -g npm@6.10.3
```
Comment (ljharb) [inline appveyor.yml]: please prefer the if-based check; this won't work on node 4, for example.

Comment (raphinesse) [reply appveyor.yml]: Node 4 is commented out in the Appveyor config. I usually only consider live code. But this branch is exploratory anyhow. I do not find the `npm` version pinning done here to be a satisfactory solution to the `nyc` problem (See https://github.com/benmosher/eslint-plugin-import/pull/1648#issuecomment-583602524).

Comment (ljharb) [reply appveyor.yml]: yes, but the intention is to uncomment it.

Fair that it's exploratory, but we were down to 1 failing test before you both started force pushing everywhere :-)

Comment (ljharb) [reply appveyor.yml]: As for local dev, the first step is to get CI working - all local dev by maintainers is done on a Mac.

Once CI is working, we can try to fix local Windows dev, for the rare contributor that uses it.

Comment (raphinesse) [reply appveyor.yml]: > yes, but the intention is to uncomment it.

Understood 


> Fair that it's exploratory, but we were down to 1 failing test before you both started force pushing everywhere :-)

But test not running at all on Node 8 and 10.

Comment (ljharb) [reply appveyor.yml]: Again, first step is to get CI passing *somehow*.

We can uncomment node 10, 8, 6, 4, etc on appveyor in a future PR.

Comment (raphinesse) [reply appveyor.yml]: > As for local dev, the first step is to get CI working - all local dev by maintainers is done on a Mac.

Understood

**Comment by raphinesse** [comment]
@ljharb I'm done force-pushing here. Can confirm that `npm` version pinning resolves the `nyc` problem on all Node versions currently running on AppVeyor.

**Comment by ljharb** [comment]
It does still have to stay open until it’s merged tho :-)
